### PR TITLE
Fix pthread_kill behaviour

### PR DESCRIPTION
**Pull request by sbc100**
All `pthread_kill` really does is ensure that the signal handler for a given signal occurs on a specific thread.

The slight exception to this is `SIGCANCEL` which is how `pthread_cancel` is implemented internally.

This allows us to completely remove the `cancelThread` and `killThread` JS functions.

Fixes: #22274

### CODE REVIEW

File: test/pthread/test_pthread_kill.c
Comment (kripken) [inline test/pthread/test_pthread_kill.c:43]: Why was this sleep added to the test?

File: system/lib/pthread/pthread_kill.c
Comment (kripken) [inline system/lib/pthread/pthread_kill.c:43]:
```suggestion
  // The job of pthread_kill is basically to run the (process-wide) signal
```

File: system/lib/pthread/pthread_kill.c
Comment (kripken) [inline system/lib/pthread/pthread_kill.c:19]:
```suggestion
    // For SIGCANCEL it is simply enough to run the proxied function since we call
```

File: system/lib/pthread/pthread_kill.c
Comment (kripken) [inline system/lib/pthread/pthread_kill.c:22]:
```suggestion
    // proxying system itself is on the stack and we want to exit with clean
```

File: system/lib/pthread/pthread_kill.c
Comment (kripken) [inline system/lib/pthread/pthread_kill.c:24]: Does "run the proxied function" refer to `_emscripten_runtime_keepalive_clear`? In `action_terminate()` in `raise.c` I see we also do `_Exit(128 + sig);`

Comment (sbc100) [reply system/lib/pthread/pthread_kill.c:24]: `SIGCANCEL` doesn't want to terminate the whole process, only the current thread.    Its a special internal signal that effects just the current thread.

File: system/lib/pthread/pthread_kill.c
Comment (sbc100) [reply system/lib/pthread/pthread_kill.c:19]: "the proxied function" refers to the signal handler setup by sigaction that raise will call.

File: system/lib/pthread/pthread_kill.c
Comment (sbc100) [reply system/lib/pthread/pthread_kill.c:22]: I updated this comment.

File: system/lib/pthread/pthread_kill.c
Comment (sbc100) [reply system/lib/pthread/pthread_kill.c:24]: I updated this comment.

File: test/pthread/test_pthread_kill.c
Comment (sbc100) [reply test/pthread/test_pthread_kill.c:43]: In order to cancel a running thread the threads needs to reach some cancellation point, in this case the cancellation point is the mailbox check where this thread runs the signal handler code.

Comment (kripken) [reply test/pthread/test_pthread_kill.c:43]: How did it work before? If it failed before, does that mean users might notice this change?

Comment (sbc100) [reply test/pthread/test_pthread_kill.c:43]: Ah yes, it is possible that this change is observable.  Previously we used `worker.terminate()` to cancel the running thread, but now we require the thread to cooperate somewhat.  It would take a very uncoperative thread indeed to observe this though.  Basically a busy loop like this.

If you prefer I can seperate this change in to just the `pthread_kill` part, and keep the `pthread_cancel` behaviour the same (i.e. keep using `worker.terminate()`).

Comment (sbc100) [reply test/pthread/test_pthread_kill.c:43]: Having said that this test is simply wrong since it was using `pthread_kill` to try to kill a thread.. since is not what that API does.   The public `pthread_kill` API will always being down the entire process.  So this test simply has the wrong expectation of how the API works.

Comment (sbc100) [reply test/pthread/test_pthread_kill.c:43]: Actually I'm wrong.. pthread_kill was implement via `worker.terminate()` but this was always wrong.  `pthread_cancel` has always requires cooperation from the target there.

Thus the only folks who could observe this change are ones (like this test) that were wrongly assuming that pthread_kill could be used to terminate single thread.

Thus, I think this change is good as is.

Comment (kripken) [reply test/pthread/test_pthread_kill.c:43]: Change sounds good, then. But maybe add a changelog mention?

**Review by kripken — APPROVED** [review]
Overall this lgtm but I am not deeply familiar with the pthreads API or this implementation code, so it might be good to wait for another review here.

**Comment by sbc100** [comment]
OK added a changlog entry.

The only part of really worried about is putting the pthread_testcancel in at the end of mailbox check.   This seem correct to me and @tlively can confirm... its not safe to unwind/exit in the middle of processing and event?

Otherwise I'm pretty confident that this change in the right direction and will land before EOD.

File: system/lib/pthread/pthread_kill.c
Comment (tlively) [inline system/lib/pthread/pthread_kill.c:20]: `The calling then` => `The calling thread`

File: system/lib/pthread/pthread_kill.c
Comment (tlively) [inline system/lib/pthread/pthread_kill.c:27]: `we return though` => `we return, though`

File: test/pthread/test_pthread_kill.c
Comment (tlively) [inline test/pthread/test_pthread_kill.c:59]: Can we use condition variables instead of sleeps to make the test run as fast as possible?

**Review by tlively — COMMENTED** [review]
Looks good overall 👍

**Comment by tlively** [comment]
> The only part of really worried about is putting the pthread_testcancel in at the end of mailbox check. This seem correct to me and @tlively can confirm... its not safe to unwind/exit in the middle of processing and event?

Right. I think unwinding in the middle of processing a mailbox item would be asking for trouble, although I don't have enough of the system paged into my memory to say exactly what would go wrong.

File: test/pthread/test_pthread_kill.c
Comment (sbc100) [reply test/pthread/test_pthread_kill.c:59]: Done: https://github.com/emscripten-core/emscripten/pull/22479

**Comment by juj** [comment]
pthread_cancel() should definitely follow well-orderedness of execution. It'll prove in the wild whether it's correct.

Before this change, the pthread code didn't really support signal API at all. Now that it does, it might be worth looking into running some of the openposixtestsuite signal API tests.

**Pull request merged**
